lib/api/apiUtils/quotas/quotaUtils.js

```diff
@@ -118,113 +247,51 @@ function validateQuotas(request, bucket, account, apiNames, apiMethod, inflight,
-                    return QuotaService.getUtilizationMetrics('account',
-                        account.account, null, {
-                        action: apiName,
-                        ..._inflights,
-                    }, (err, accountMetrics) => {
-                        if (err) {
-                            return parallelDone(err);
-                        }
-                        if (!isMetricStale(accountMetrics, 'account', account.account, apiName, inflight, log) &&
-                            accountMetrics.bytesTotal + _inflightForCheck > accountQuota) {
-                            log.debug('Account quota exceeded', {
-                                accountId: account.account,
-                                action: apiName,
-                                inflight,
-                                quota: accountQuota,
-                                bytesTotal: accountMetrics.bytesTotal,
-                            });
-                            accountQuotaExceeded = true;
-                        }
-                        return parallelDone();
-                    });
+
+                bucketQuotaExceeded = _bucketQuotaExceeded;
+                accountQuotaExceeded = _accountQuotaExceeded;
+
+                // Inflights are inverted: in case of cleanup, we just re-issue
+                // the same API call.
+                if (_inflights) {
+                    _inflights = -_inflights;
                 }
-                return parallelDone();
-            },
-        }, err => {
-            if (err) {
-                return done(err);
-            }
-            // Inflights are inverted: in case of cleanup, we just re-issue
-            // the same API call.
-            if (_inflights) {
-                _inflights.inflight = -_inflights.inflight;
-            }
-            const removeInflights = (errorFromAPI, _done) => {
-                // Quotas are cleaned only in case of error in the API
-                if (!errorFromAPI) {
+
+                request.finalizerHooks.push((errorFromAPI, _done) => {
+                    const code = (bucketQuotaExceeded || accountQuotaExceeded) ? 429 : 200;
+                    const quotaCleanUpStartTime = process.hrtime.bigint();
+                    // Quotas are cleaned only in case of error in the API
+                    if (errorFromAPI) {
+                        return _evaluateQuotas(bucketQuota, accountQuota, bucket, account, _inflights,
+                            null, apiName, true, log, () => {
+                                monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration +
+                                    Number(process.hrtime.bigint() - quotaCleanUpStartTime));
+                                return _done();
+                            });
+                    }
+                    monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration);
                     return _done();
```

Comment:
use waterflow to simplify code:

```suggestion
                    async.waterflow(
                        cb => {
                            if (errorFromAPI) {
                                return _evaluateQuotas(bucketQuota,      accountQuota, bucket, account, _inflights,
                            null, apiName, true, log, cb);
                            }
                            return cb();
                        },
                        () => {
                            monitorQuotaEvaluationDuration(apiMethod, type, code, quotaEvaluationDuration +
                                Number(process.hrtime.bigint() - quotaCleanUpStartTime));
                            return _done();
                        }
                    )
```